storage/common/include/algebra/storage/matrix.hpp

```diff
@@ -0,0 +1,454 @@
+/** Algebra plugins, part of the ACTS project
+ *
+ * (c) 2024 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+#pragma once
+
+// @TODO: Remove this when Vc fixes their false positives.
+#if defined(__GNUC__) && !defined(__clang__)
+#pragma GCC diagnostic warning "-Wdeprecated-declarations"
+#endif
+
+// Project include(s).
+#include "algebra/storage/vector.hpp"
+
+// System include(s).
+#include <array>
+#include <cassert>
+
+namespace algebra::storage {
+
+/// Generic matrix type that can take vectors as columns
+template <template <typename, std::size_t> class array_t, typename value_t,
+          std::size_t ROW, std::size_t COL>
+struct alignas(alignof(storage::vector<ROW, value_t, array_t>)) matrix {
+
+  // The matrix consists of column vectors
+  using vector_type = storage::vector<ROW, value_t, array_t>;
+  // Value type: Can be simd types
+  using value_type = value_t;
+
+  /// Default constructor
+  constexpr matrix() = default;
```

Comment:
Yay on the `constexpr`. But I would still use `ALGEBRA_HOST_AND_DEVICE`. So that clients would not be forced into using a (still) experimental feature with CUDA.

Reply:
If I use ```ALGEBRA_HOST_AND_DEVICE``` with a defaulted constructor CUDA will complain

Reply:
I should've opened the comment on a different function. (You're right, on this particular function you mustn't add the macro.)

But on all the other functions, I feel pretty strongly about `ALGEBRA_...`. Since this is in a common piece of code here. We should make it technically allowed to use this code on a GPU.

Reply:
Some of this code uses vectorization specific functionality like boolean mask evaluation. Maybe we can hide that behind an interface such that it has a chance of ever compiling in device code?